packages/icon/src/vaadin-iconset-mixin.js

```diff
@@ -123,23 +100,75 @@ export const IconsetMixin = (superClass) =>
         initIconsMap(iconset, name);
         iconset.size = size;
         iconset.name = name;
-
-        // Call this function manually instead of using observer
-        // to make it work without appending element to the DOM.
-        iconset.__nameChanged(name);
       }
     }
 
+    /**
+     * The name of the iconset. Every iconset is required to have its own unique name.
+     * All the SVG icons in the iconset must have IDs conforming to its name.
+     *
+     * See also [`name`](#/elements/vaadin-icon#property-name) property of `vaadin-icon`.
+     *
+     * @return {string}
+     */
+    get name() {
+      return this.__name;
+    }
+
+    /**
+     * @type {string}
+     */
+    set name(name) {
+      const oldName = this.__name;
+      this.__name = name;
+      this.__nameChanged(name, oldName);
+    }
+
+    /**
+     * The size of an individual icon. Note that icons must be square.
+     *
+     * When using `vaadin-icon`, the size of the iconset will take precedence
+     * over the size defined by the user to ensure correct appearance.
+     *
+     * @return {number}
+     */
+    get size() {
+      // Use default property value as a fallback here instead of the constructor
+      // to not override an instance property in the lazy upgrade scenario below.
+      return this.__size !== undefined ? this.__size : 24;
+    }
+
+    /**
+     * @type {number}
+     */
+    set size(size) {
+      this.__size = size;
+    }
+
     /** @protected */
     connectedCallback() {
-      super.connectedCallback();
+      // A user may set a property on an _instance_ of an element
+      // before the custom element is lazily imported and upgraded.
+      // If so, we need to run it through the proper class setter.
+      ['name', 'size'].forEach((prop) => {
+        // eslint-disable-next-line no-prototype-builtins
+        if (this.hasOwnProperty(prop)) {
+          const value = this[prop];
+          delete this[prop];
+          this[prop] = value;
+        }
+      });
+
       this.style.display = 'none';
+    }
 
-      // Store reference and init icons.
-      const { name } = this;
-      iconsetRegistry[name] = this;
-      initIconsMap(this, name);
-      this.__updateIcons(name);
```

Comment:
Moved this logic to `__nameChanged` method since it's now called when `name` property is set and is no longer an "observer" (observers are only initialized when the element is connected for the first time).